flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogMigrationRestoreTarget.java

```diff
@@ -231,8 +239,15 @@ public <N, SV, SEV, S extends State, IS extends S> IS createInternalState(
                             StateSnapshotTransformer.StateSnapshotTransformFactory<SEV>
                                     snapshotTransformFactory)
                     throws Exception {
-                return keyedStateBackend.createInternalState(
-                        namespaceSerializer, stateDesc, snapshotTransformFactory);
+                ChangelogState changelogState =
+                        changelogStateFactory.getExistingState(
+                                stateDesc.getName(),
+                                StateMetaInfoSnapshot.BackendStateType.KEY_VALUE);
+                return changelogState == null
+                        ? keyedStateBackend.createInternalState(
+                                namespaceSerializer, stateDesc, snapshotTransformFactory)
+                        : keyedStateBackend.upgradeKeyedState(
+                                namespaceSerializer, stateDesc, snapshotTransformFactory);
```

Comment:
Could you explain why we need to call `upgradeKeyedState` here?
Isn't this a call by a user program allready? (so no further metadata updates are expected)

Reply:
It should always be `createInternalState`. Modified.